### PR TITLE
Also use stable name in SendJoinResponse struct

### PR DESCRIPTION
**Pull request by DMRobertson**
follow-up to #14832

Sean [pointed out](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$SdA3tdegQxt8tOFAKwNPZGKLgHrNzbko-D0KpmTSrVc?via=matrix.org&via=element.io&via=beeper.com) that the response struct should mirror the spec, which I agree with. I haven't chased this through the rest of the source tree though---I don't think it'd be worth it.

### CODE REVIEW

**Review by squahtx — COMMENTED** [review]
Looks good!

I'm a little confused as to the difference between `SendJoinResult` and `SendJoinResponse`.
Should we do the rename in `SendJoinResult` too or is that sufficiently distanced from the spec that we should leave it as is?

**Comment by DMRobertson** [comment]
> I'm a little confused as to the difference between `SendJoinResult` and `SendJoinResponse`.

This also confused me. My guess is that the `Response` represents the JSON body of the result; the `Result` represents a validated, post-processed version which is only constructed after [a bunch of checks](https://github.com/matrix-org/synapse/blob/ef059a88b3d2c17b12c484bb2293892e94c1e665/synapse/federation/federation_client.py#L1042-L1148) have been, err, checked.

(I also note that `Response` has `event_dict` and `event` fields; `Result` only has `result`.)

I think the Response should be considered internal to synapse.federation.transport, and the result should be considered a public export of synapse.federation(.federation_client).

I can write this as a docstring somewhere if you think this is plausible enough.

**Comment by DMRobertson** [comment]
> Should we do the rename in `SendJoinResult` too or is that sufficiently distanced from the spec that we should leave it as is?

I'd be inclined to leave it as is, personally.

File: synapse/federation/federation_client.py [hunk 1]
```diff
@@ -1142,7 +1142,7 @@ async def _execute(pdu: EventBase) -> None:
                     % (auth_chain_create_events,)
                 )
 
-            if response.partial_state and not response.servers_in_room:
+            if response.members_omitted and not response.servers_in_room:
                 raise InvalidResponseError(
                     "partial_state was set, but no servers were listed in the room"
```
Comment (squahtx) [inline synapse/federation/federation_client.py:1147]:
```suggestion
                    "members_omitted was set, but no servers were listed in the room"
```

**Review by squahtx — APPROVED** [review]
lgtm!

File: synapse/federation/federation_client.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/federation/federation_client.py:1147]: Argh, I'm blind. Thanks

**Comment by DMRobertson** [comment]
There's a complement failure which I assume is a flake (#14853). Rerunning...

**Pull request merged**